File: sources/iOS/LuaMobileSdk/LuaMobileSdk/Sources/Api/engine/register/LuaPlainDataClassRegister.m
/**
 * Android LuaMobileSdk for iOS framework project.
 *
 * Copyright 2017 Arno Zhang <zyfgood12@163.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "LuaPlainDataClassRegister.h"
#import "LuaState.h"
#import "LuaDeclares.h"
#import "NSString+Sdk.h"
#import "LuaBridgeUtils.h"
#import "SdkDeclares.h"


int sdk_PlainClassLenInvoker(lua_State* ls) {
    MethodInvokeWrapper* wrapper = sdk_getInvokerWrapper(ls);
    if (!wrapper) {
        return 0;
    }

    LuaState* state = (__bridge LuaState*) wrapper->luaState;
    LuaMethodInvokeEnv* env = (__bridge LuaMethodInvokeEnv*) wrapper->env;
    LuaPlainDataBridge* bridge = (LuaPlainDataBridge*) env.bridge;

    // a.__index = function(object) ... end
    //                         1
    //
    NSObject* object = [state toObject:1];
    [state pushNumber:[bridge length:object]];

    return 1;
}

int sdk_PlainClassIndexInvoker(lua_State* ls) {
    MethodInvokeWrapper* wrapper = sdk_getInvokerWrapper(ls);
    if (!wrapper) {
        return 0;
    }

    LuaState* state = (__bridge LuaState*) wrapper->luaState;
    LuaMethodInvokeEnv* env = (__bridge LuaMethodInvokeEnv*) wrapper->env;
    LuaPlainDataBridge* bridge = (LuaPlainDataBridge*) env.bridge;

    // a.__index = function(object, name) ... end
    //                         1     2
    //
    NSObject* object = [state toObject:1];
    NSString* name = [state toString:2];
    [state push:[bridge getData:object Name:name]];

    return 1;
}

int sdk_PlainClassNewIndexInvoker(lua_State* ls) {
    MethodInvokeWrapper* wrapper = sdk_getInvokerWrapper(ls);
    if (!wrapper) {
        return 0;
    }

    LuaState* state = (__bridge LuaState*) wrapper->luaState;
    LuaMethodInvokeEnv* env = (__bridge LuaMethodInvokeEnv*) wrapper->env;
    LuaPlainDataBridge* bridge = (LuaPlainDataBridge*) env.bridge;

    // a.__index = function(object, name, value) ... end
    //                         1     2     3
    //
    NSObject* object = [state toObject:1];
    NSString* name = [state toString:2];
    NSObject* value = [state toObjectFixed:3];
    [bridge setData:object Name:name Value:value];

    return 0;
}

@implementation LuaPlainDataClassRegister


- (void)destroy {
    [super destroy];
}

- (void)registerPlainDataClass:(LuaPlainDataBridge*)bridge {
    if (![super registerClassPrepare:bridge]) {
        return;
    }

    [mLuaState pop:2];
}

- (void)bindMetaMethods:(LuaClassBridge*)bridge {
    [super bindMetaMethods:bridge];

    // meta.__len = len
    [self newWrappedClosureAtTop:bridge Method:nil Closure:&sdk_PlainClassLenInvoker];
    [mLuaState setField:-2 CStrKey:__len];

    // meta.__index = index
    [self newWrappedClosureAtTop:bridge Method:nil Closure:&sdk_PlainClassIndexInvoker];
    [mLuaState setField:-2 CStrKey:__index];

    // meta.__newindex = newindex
    [self newWrappedClosureAtTop:bridge Method:nil Closure:&sdk_PlainClassNewIndexInvoker];
    [mLuaState setField:-2 CStrKey:__newindex];
}

- (LuaBridgedFunction*)createLenFunction:(LuaPlainDataBridge*)bridge {
    sdk_weakify(bridge);
    return [[LuaBridgedFunction alloc]
            initWith:mScriptEngine Invoker:^NSObject*(LuaScriptEngine* engine, LuaBridgeArgs* args) {
                sdk_strongify(bridge);

                if (![args atLeastHas:1]) {
                    return @(0);
                }

                NSObject* object = [args toObject:0];
                return @([bridge length:object]);
            }];
}

- (LuaBridgedFunction*)createIndexFunction:(LuaPlainDataBridge*)bridge {
    sdk_weakify(bridge);
    return [[LuaBridgedFunction alloc]
            initWith:mScriptEngine Invoker:^NSObject*(LuaScriptEngine* engine, LuaBridgeArgs* args) {
                sdk_strongify(bridge);

                if (![args atLeastHas:2]) {
                    return nil;
                }

                NSString* name = [args toString:1];
                if (!name || [name sdk_isEmpty]) {
                    return nil;
                }

                NSObject* object = [args toObject:0];
                return [bridge getData:object Name:name];
            }];
}

- (LuaBridgedFunction*)createNewIndexFunction:(LuaPlainDataBridge*)bridge {
    sdk_weakify(bridge);
    return [[LuaBridgedFunction alloc]
            initWith:mScriptEngine Invoker:^NSObject*(LuaScriptEngine* engine, LuaBridgeArgs* args) {
                sdk_strongify(bridge);

                if (![args atLeastHas:3]) {
                    return nil;
                }

                NSString* name = [args toString:1];
                if (!name || [name sdk_isEmpty]) {
                    return nil;
                }

                NSObject* object = [args toObject:0];
                NSObject* value = [args toObjectFixed:2];
                [bridge setData:object Name:name Value:value];
                return nil;
            }];
}

@end

File: sources/Android/sdk/src/main/java/com/lua/mobile/sdk/engine/register/LuaPlainDataClassRegister.java
/**
 * Android LuaMobileSdk for Android framework project.
 *
 * Copyright 2016 Arno Zhang <zyfgood12@163.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lua.mobile.sdk.engine.register;

import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.util.Log;
import com.lua.mobile.sdk.engine.LuaScriptEngine;
import com.lua.mobile.sdk.engine.args.LuaBridgeArgs;
import com.lua.mobile.sdk.engine.args.LuaPlainDataBridgeArgs;
import com.lua.mobile.sdk.engine.bridge.LuaBridgedFunction;
import com.lua.mobile.sdk.engine.bridge.LuaClassBridge;
import com.lua.mobile.sdk.engine.bridge.LuaPlainDataBridge;
import com.lua.mobile.sdk.engine.lua.LuaDeclares;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;

public class LuaPlainDataClassRegister extends LuaAbstractClassRegister {

    private static final String TAG = "LuaPlainDataClassRegister";


    public LuaPlainDataClassRegister(LuaScriptEngine scriptEngine) {
        super(scriptEngine);
    }

    /**
     * _G.ClassName = bridge
     *
     * a = ClassName()
     * #a                       ->      a.meta.__len(a)
     * local value = a.field    ->      a.meta.__index(a, field)
     * a.field = 10             ->      a.meta.__newindex(a, field, 10)
     */
    public <T> void registerPlainDataClass(@NonNull final LuaPlainDataBridge<T> bridge) {
        super.registerClassPrepare(bridge);
        mLuaState.pop(2);
    }

    @Override
    protected <T> void bindMetaMethods(int tableIndex, @NonNull LuaClassBridge<T> bridge) {
        super.bindMetaMethods(tableIndex, bridge);

        LuaPlainDataBridge<T> plainData = (LuaPlainDataBridge<T>) bridge;
        String name = bridge.getBridgeName();

        // meta.__java_meta_funcs.__len = index
        LuaBridgedFunction len = createLenFunction(plainData);
        len.setFunctionName(name + "." + LuaDeclares.__len);
        mLuaState.pushBridgedFunction(len);
        mLuaState.setField(tableIndex, LuaDeclares.__len);

        // meta.__java_meta_funcs.__index = index
        LuaBridgedFunction index = createIndexFunction(plainData);
        index.setFunctionName(name + "." + LuaDeclares.__index);
        mLuaState.pushBridgedFunction(index);
        mLuaState.setField(tableIndex, LuaDeclares.__index);

        // meta.__java_meta_funcs.__newindex = newIndex
        LuaBridgedFunction newIndex = createNewIndexFunction(plainData);
        newIndex.setFunctionName(name + "." + LuaDeclares.__newindex);
        mLuaState.pushBridgedFunction(newIndex);
        mLuaState.setField(tableIndex, LuaDeclares.__newindex);
    }

    @Override
    protected void afterBindMetaMethods(int tableIndex) {
        super.afterBindMetaMethods(tableIndex);

        mLuaState.bindMetaLen(tableIndex);
        mLuaState.bindMetaIndex(tableIndex);
        mLuaState.bindMetaNewIndex(tableIndex);
    }

    private <T> LuaBridgedFunction createLenFunction(@NonNull final LuaPlainDataBridge<T> bridge) {
        return new LuaBridgedFunction(mScriptEngine) {

            @Override
            protected LuaBridgeArgs createBridgeArgs() {
                return new LuaPlainDataBridgeArgs(mScriptEngine);
            }

            @Override
            protected int invokeInternal(LuaBridgeArgs args) throws Throwable {
                if (!args.atLeastHas(1)) {
                    mLuaState.push(0);
                    return 1;
                }

                T object = (T) args.toJavaObject(0);
                mLuaState.push(bridge.length(object));

                return 1;
            }
        };
    }

    private <T> LuaBridgedFunction createIndexFunction(@NonNull final LuaPlainDataBridge<T> bridge) {
        return new LuaBridgedFunction(mScriptEngine) {

            @Override
            protected LuaBridgeArgs createBridgeArgs() {
                return new LuaPlainDataBridgeArgs(mScriptEngine);
            }

            @Override
            protected int invokeInternal(LuaBridgeArgs args) throws Throwable {
                if (!args.atLeastHas(2)) {
                    mLuaState.pushNil();
                    return 1;
                }

                String name = args.toString(1);
                if (TextUtils.isEmpty(name)) {
                    mLuaState.pushNil();
                    return 1;
                }

                Object data = null;
                T object = (T) args.toJavaObject(0);

                if (bridge.accessClassField()) {
                    Field field = null;
                    try {
                        field = bridge.getDataClass().getField(name);
                        if (field != null) {
                            int modifiers = field.getModifiers();
                            if (Modifier.isPublic(modifiers)) {
                                data = field.get(object);
                            }
                        }
                    } catch (Exception e) {
                        Log.e(TAG, String.format("cannot access __index. class: %s, name: %s",
                                bridge.getDataClass().getSimpleName(), name), e);
                    }
                }

                if (data == null) {
                    data = bridge.getData(object, name);
                }

                if (data == null) {
                    mLuaState.pushNil();
                    return 1;
                }

                mLuaState.push(data);
                return 1;
            }
        };
    }

    private <T> LuaBridgedFunction createNewIndexFunction(@NonNull final LuaPlainDataBridge<T> bridge) {
        return new LuaBridgedFunction(mScriptEngine) {

            @Override
            protected LuaBridgeArgs createBridgeArgs() {
                return new LuaPlainDataBridgeArgs(mScriptEngine);
            }

            @Override
            protected int invokeInternal(LuaBridgeArgs args) throws Throwable {
                if (!args.atLeastHas(2)) {
                    return 0;
                }

                String name = args.toString(1);
                if (TextUtils.isEmpty(name)) {
                    mLuaState.pushNil();
                    return 1;
                }

                T object = (T) args.toJavaObject(0);
                Object data = null;
                if (args.atLeastHas(3)) {
                    data = args.toJavaObject(2);
                }

                if (bridge.accessClassField()) {
                    try {
                        Field field = bridge.getDataClass().getField(name);
                        if (field != null) {
                            int modifiers = field.getModifiers();
                            if (Modifier.isPublic(modifiers)) {
                                field.set(object, data);
                                return 0;
                            }
                        }
                    } catch (Exception e) {
                        Log.e(TAG, String.format("cannot access __newindex. class: %s, name: %s, value: %s",
                                bridge.getDataClass().getSimpleName(), name, String.valueOf(data)), e);
                    }
                }

                bridge.setData(object, name, data);
                return 0;
            }
        };
    }
}
